src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -0,0 +1,2966 @@
+            comp.VerifyEmitDiagnostics(
+                // (13,24): error CS9500: Cannot initialize type 'S<object>?' with a collection literal because the type is not constructible.
+                //         S<object>? x = [];
+                Diagnostic(ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, "[]").WithArguments("S<object>?").WithLocation(13, 24),
+                // (14,13): error CS9500: Cannot initialize type 'S<object>?' with a collection literal because the type is not constructible.
+                //         x = [];
+                Diagnostic(ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, "[]").WithArguments("S<object>?").WithLocation(14, 13),
+                // (15,24): error CS9500: Cannot initialize type 'S<object>?' with a collection literal because the type is not constructible.
+                //         S<object>? y = [1];
+                Diagnostic(ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, "[1]").WithArguments("S<object>?").WithLocation(15, 24),
+                // (16,13): error CS9500: Cannot initialize type 'S<object>?' with a collection literal because the type is not constructible.
+                //         y = [2];
+                Diagnostic(ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, "[2]").WithArguments("S<object>?").WithLocation(16, 13));
+        }
+
+        [Fact]
+        public void OrderOfEvaluation()
+        {
+            string source = """
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                class C<T> : IEnumerable
+                {
+                    private List<T> _list = new List<T>();
+                    public void Add(T t)
+                    {
+                        Console.WriteLine("Add {0}", t);
+                        _list.Add(t);
+                    }
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        C<int> x = [Get(1), Get(2)];
+                        C<C<int>> y = [[Get(3)], [Get(4), Get(5)]];
+                    }
+                    static int Get(int value)
+                    {
+                        Console.WriteLine("Get {0}", value);
+                        return value;
+                    }
+                }
+                """;
+            CompileAndVerify(source, expectedOutput: """
+                Get 1
+                Add 1
+                Get 2
+                Add 2
+                Get 3
+                Add 3
+                Add C`1[System.Int32]
+                Get 4
+                Add 4
+                Get 5
+                Add 5
+                Add C`1[System.Int32]
+                """);
+        }
+
+        // Ensure collection literal conversions are not standard implicit conversions
```

Comment:
Is there discussion I can review which explains why we want this?

Reply:
The only discussion was in https://github.com/dotnet/roslyn/pull/67025#discussion_r1122325444. I have added an open question to the test plan https://github.com/dotnet/roslyn/issues/66418.